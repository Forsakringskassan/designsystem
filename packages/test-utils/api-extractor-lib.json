{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
