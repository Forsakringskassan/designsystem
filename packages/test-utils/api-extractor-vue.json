{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
