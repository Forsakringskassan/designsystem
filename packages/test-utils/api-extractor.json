{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
