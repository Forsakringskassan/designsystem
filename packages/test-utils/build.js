const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
