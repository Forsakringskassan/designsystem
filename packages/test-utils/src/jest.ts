/* imported for side-effects only */
import "./matchers";

export * from "./lib";
