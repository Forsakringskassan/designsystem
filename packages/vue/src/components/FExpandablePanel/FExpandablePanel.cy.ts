import { type DefineComponent, defineComponent } from "vue";
import {
    DensityWrapper,
    densityWrapperHeight,
    densityWrapperWidth,
} from "@fkui/test-utils/vue";
import { FExpandablePanelPageObject } from "../../cypress";
import FExpandablePanel from "./FExpandablePanel.vue";

function createComponent(template: string): DefineComponent {
    return defineComponent({
        template,
        components: { FExpandablePanel },
        data() {
            return {
                expanded: false,
                type: Boolean,
            };
        },
        methods: {
            onToggle() {
                this.expanded = !this.expanded;
            },
        },
    });
}

describe("FExpandablePanel", () => {
    const panel = new FExpandablePanelPageObject(
        "[data-test=expandable-panel]",
    );

    const panelWithNotification = new FExpandablePanelPageObject(
        "[data-test=notification-example] .expandable-panel",
    );

    beforeEach(() => {
        cy.clearLocalStorage();
    });

    it("Should have a page object that can access any necessary elements for default expandable panel", () => {
        const template = /* HTML */ `
            <f-expandable-panel
                :expanded="expanded"
                @toggle="onToggle"
                v-test="'expandable-panel'"
            >
                <template #title> Titel </template>
                <template #default>
                    Innehåll
                    <p>
                        <a class="anchor" href="" target="_blank">
                            Länk till annan sida
                        </a>
                    </p>
                </template>
            </f-expandable-panel>
        `;
        cy.mount(createComponent(template));
        panel.el().should("be.visible");
        panel.isOpen().should("be.false");
        panel.expandCollapseIcon().click();
        panel.isOpen().should("be.true");
        panel.header().should("have.trimmedText", "Titel");
        panel.notificationIcon().should("not.exist");
    });
    it("Should have a page object that can access any necessary elements for expandable panel with notification icon", () => {
        const template = /* HTML */ `
            <div v-test="'notification-example'">
                <f-expandable-panel
                    :expanded="expanded1"
                    :notifications="1"
                    @toggle="onToggle1"
                >
                    <template #title> Titel med en notifiering </template>
                    <template #default> Innehåll </template>
                    <template #outside>
                        Relaterat innehåll som visas när panelen är expanderad
                        men utanför body
                    </template>
                </f-expandable-panel>
            </div>
        `;
        cy.mount(createComponent(template));
        panelWithNotification.notificationIcon().should("be.visible");
        panelWithNotification
            .notificationIcon()
            .should("have.trimmedText", "Du har 1 notifieringar.");
        panelWithNotification.header().click();
        panelWithNotification.body().should("have.trimmedText", "Innehåll");
        panelWithNotification
            .relatedInfo()
            .should(
                "have.trimmedText",
                "Relaterat innehåll som visas när panelen är expanderad men utanför body",
            );

        panelWithNotification.numberOfNotifications().should("be.equal", 1);
    });

    /* eslint-disable-next-line mocha/no-skipped-tests -- temporary to get builds running */
    describe.skip("density", () => {
        const DensityComponent = defineComponent({
            template: /* HTML */ `
                <density-wrapper>
                    <f-expandable-panel :expanded="true">
                        <template #title> Rubrik </template>
                        <template #default> Innehåll </template>
                    </f-expandable-panel>
                </density-wrapper>
            `,
            components: {
                DensityWrapper,
                FExpandablePanel,
            },
        });

        it(`should be densified`, () => {
            cy.viewport(densityWrapperWidth, densityWrapperHeight);
            cy.mount(DensityComponent);
            cy.toMatchScreenshot();
        });
    });
});
