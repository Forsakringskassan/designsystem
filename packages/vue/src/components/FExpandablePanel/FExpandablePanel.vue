<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService } from "@fkui/logic";
import { FIcon } from "../FIcon";
import { FExpand } from "../FExpand";
import { hasSlot } from "../../utils";
import { TranslationMixin } from "../../plugins";

export default defineComponent({
    name: "FExpandablePanel",
    components: {
        FIcon,
        FExpand,
    },
    mixins: [TranslationMixin],
    inheritAttrs: false,
    props: {
        /**
         * Toggle expanded/collapsed state
         */
        expanded: {
            type: Boolean,
            required: false,
            default: false,
        },
        /**
         * Element to render for the header element inside the expandable panel.
         * 'h1', 'h2', 'h3', 'h4', 'h5' and 'h6' is valid.
         */
        headerTag: {
            default: "h2",
            required: false,
            validator(value: string): boolean {
                return ["h1", "h2", "h3", "h4", "h5", "h6"].includes(value);
            },
        },
        /**
         * The id for the content id attribute.
         * If the prop is not set the id will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },

        /**
         * Number of notifications present in panel.
         *
         * If set to zero (default) no notification badge will be displayed.
         */
        notifications: {
            type: Number,
            required: false,
            default: 0,
        },
        /**
         * Text template for the screen reader notification text.
         * Use `%VALUE%` as a placeholder for the number of notifications.
         */
        screenReaderNotificationTemplate: {
            type: String,
            required: false,
            default: "Du har %VALUE% notifieringar.",
        },
    },
    emits: ["toggle"],
    computed: {
        expandedClass(): string {
            return this.expanded ? "expandable-panel--expanded" : "expandable-panel--collapsed";
        },
        hasOutsideSlot(): boolean {
            return hasSlot(this, "outside");
        },
        haveNotifications(): boolean {
            return this.notifications > 0;
        },
        screenReaderNotificationText(): string {
            return `${this.screenReaderNotificationTemplate.replace("%VALUE%", this.notifications.toString())}`;
        },
    },
    methods: {
        onClickHeadingButton(event: MouseEvent): void {
            /**
             * Emitted when panel heading is clicked.
             * @type {MouseEvent}
             */
            this.$emit("toggle", event);
        },
    },
});
</script>

<template>
    <div class="expandable-panel" :class="expandedClass">
        <component :is="headerTag" class="expandable-panel__heading">
            <button
                type="button"
                :aria-expanded="expanded ? 'true' : 'false'"
                :aria-controls="id"
                v-bind="$attrs"
                @click="onClickHeadingButton"
            >
                <span class="expandable-panel__icon">
                    <span class="icon-stack">
                        <f-icon name="dash"></f-icon>
                        <f-icon name="dash"></f-icon>
                    </span>
                </span>

                <!-- @slot Slot used for title content -->
                <slot name="title"></slot>

                <span
                    v-if="haveNotifications"
                    class="expandable-panel__notification"
                    :title="
                        $t('fkui.expandable-panel.notification.title', '{{ count }} notifiering{{ suffix }}', {
                            count: notifications,
                            suffix: notifications > 1 ? 'ar' : '',
                        })
                    "
                >
                    <span class="sr-only">{{ screenReaderNotificationText }}</span>
                    <f-icon name="bell"></f-icon>
                </span>
            </button>
        </component>

        <f-expand>
            <div v-show="expanded" :id="id" class="expandable-panel__content">
                <div class="expandable-panel__body">
                    <!-- @slot Slot used for content shown when panel is expanded -->
                    <slot></slot>
                </div>
                <div v-if="hasOutsideSlot" class="expandable-panel__outside">
                    <!-- @slot Slot used for optional content rendered below panel when expanded -->
                    <slot name="outside"></slot>
                </div>
            </div>
        </f-expand>
    </div>
</template>
