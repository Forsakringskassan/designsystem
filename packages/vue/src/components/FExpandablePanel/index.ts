export { default as FExpandablePanel } from "./FExpandablePanel.vue";
