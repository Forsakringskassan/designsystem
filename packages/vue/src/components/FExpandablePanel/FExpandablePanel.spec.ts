import { VueWrapper, mount } from "@vue/test-utils";
import { type ConfigData } from "html-validate";
import FExpandablePanel from "./FExpandablePanel.vue";
import "html-validate/jest";

const config: ConfigData = {
    root: true,
    rules: {
        "no-inline-style": "off",
    },
};

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FExpandablePanel, {
        attrs: { ...attrs },
        props: { id: "my-id", ...props },
        slots: { title: "My panel title", body: "Lorem ipsum", ...slots },
        global: {
            stubs: ["f-expand", "f-icon"],
        },
    });
}

describe("snapshots", () => {
    it("should match snapshot when collapsed", () => {
        expect.assertions(2);
        const wrapper = createWrapper();
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });

    it("should match snapshot when expanded", () => {
        expect.assertions(2);
        const wrapper = createWrapper({ props: { expanded: true } });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });

    it("should match snapshot with notification", () => {
        expect.assertions(2);
        const wrapper = createWrapper({ props: { notifications: 2 } });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });

    it("should match snapshot with generated id", () => {
        expect.assertions(2);
        const wrapper = createWrapper({ props: { id: undefined } });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });

    it("should match snapshot with custom heading level", () => {
        expect.assertions(2);
        const wrapper = createWrapper({ props: { headerTag: "h3" } });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });

    it('should match snapshot with "outside" slot', () => {
        expect.assertions(2);
        const wrapper = createWrapper({ slots: { outside: "dolor sit amet" } });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.element).toHTMLValidate(config);
    });
});

describe("attributes", () => {
    it("should pass attributes", () => {
        const wrapper = createWrapper({
            attrs: {
                disabled: true,
            },
        });

        const input = wrapper.get("button");

        expect(input.attributes("disabled")).toBeDefined();
    });
});

describe("events", () => {
    it("should emit toggle event", async () => {
        expect.assertions(1);
        const toggle = jest.fn();

        const wrapper = createWrapper({
            attrs: { onToggle: toggle },
        });
        const input = wrapper.get("button");
        await input.trigger("click");

        expect(toggle).toHaveBeenCalled();
    });
    it("should pass listeners", async () => {
        const foobar = jest.fn();
        const wrapper = createWrapper({
            attrs: { onFoobar: foobar },
        });
        const element = wrapper.get("button");
        await element.trigger("foobar");
        expect(foobar).toHaveBeenCalled();
    });
});
