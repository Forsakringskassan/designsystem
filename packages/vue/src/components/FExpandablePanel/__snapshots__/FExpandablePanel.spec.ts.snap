// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot when collapsed 1`] = `
<div
  class="expandable-panel expandable-panel--collapsed"
>
  <h2
    class="expandable-panel__heading"
  >
    <button
      aria-controls="my-id"
      aria-expanded="false"
      type="button"
    >
      <span
        class="expandable-panel__icon"
      >
        <span
          class="icon-stack"
        >
          <f-icon-stub
            library="f"
            name="dash"
          />
          <f-icon-stub
            library="f"
            name="dash"
          />
        </span>
      </span>
      
      My panel title
      
      <!--v-if-->
    </button>
  </h2>
  <f-expand-stub />
</div>
`;

exports[`snapshots should match snapshot when expanded 1`] = `
<div
  class="expandable-panel expandable-panel--expanded"
>
  <h2
    class="expandable-panel__heading"
  >
    <button
      aria-controls="my-id"
      aria-expanded="true"
      type="button"
    >
      <span
        class="expandable-panel__icon"
      >
        <span
          class="icon-stack"
        >
          <f-icon-stub
            library="f"
            name="dash"
          />
          <f-icon-stub
            library="f"
            name="dash"
          />
        </span>
      </span>
      
      My panel title
      
      <!--v-if-->
    </button>
  </h2>
  <f-expand-stub />
</div>
`;

exports[`snapshots should match snapshot with "outside" slot 1`] = `
<div class="expandable-panel expandable-panel--collapsed">
  <h2 class="expandable-panel__heading"><button type="button" aria-expanded="false" aria-controls="my-id"><span class="expandable-panel__icon"><span class="icon-stack"><f-icon-stub name="dash" library="f"></f-icon-stub><f-icon-stub name="dash" library="f"></f-icon-stub></span></span>My panel title
      <!--v-if-->
    </button></h2>
  <f-expand-stub></f-expand-stub>
</div>
`;

exports[`snapshots should match snapshot with custom heading level 1`] = `
<div class="expandable-panel expandable-panel--collapsed">
  <h3 class="expandable-panel__heading"><button type="button" aria-expanded="false" aria-controls="my-id"><span class="expandable-panel__icon"><span class="icon-stack"><f-icon-stub name="dash" library="f"></f-icon-stub><f-icon-stub name="dash" library="f"></f-icon-stub></span></span>My panel title
      <!--v-if-->
    </button></h3>
  <f-expand-stub></f-expand-stub>
</div>
`;

exports[`snapshots should match snapshot with generated id 1`] = `
<div class="expandable-panel expandable-panel--collapsed">
  <h2 class="expandable-panel__heading"><button type="button" aria-expanded="false" aria-controls="fkui-vue-element-0001"><span class="expandable-panel__icon"><span class="icon-stack"><f-icon-stub name="dash" library="f"></f-icon-stub><f-icon-stub name="dash" library="f"></f-icon-stub></span></span>My panel title
      <!--v-if-->
    </button></h2>
  <f-expand-stub></f-expand-stub>
</div>
`;

exports[`snapshots should match snapshot with notification 1`] = `
<div
  class="expandable-panel expandable-panel--collapsed"
>
  <h2
    class="expandable-panel__heading"
  >
    <button
      aria-controls="my-id"
      aria-expanded="false"
      type="button"
    >
      <span
        class="expandable-panel__icon"
      >
        <span
          class="icon-stack"
        >
          <f-icon-stub
            library="f"
            name="dash"
          />
          <f-icon-stub
            library="f"
            name="dash"
          />
        </span>
      </span>
      
      My panel title
      
      <span
        class="expandable-panel__notification"
        title="2 notifieringar"
      >
        <span
          class="sr-only"
        >
          Du har 2 notifieringar.
        </span>
        <f-icon-stub
          library="f"
          name="bell"
        />
      </span>
    </button>
  </h2>
  <f-expand-stub />
</div>
`;
